### PR TITLE
Add permissions anatations for fields   

### PR DESCRIPTION
**Pull request by burov**
Add permissions anatations for field, now we can set up access to field as a RO(Read only access), CO(field accepted only with create operation), add tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,6 +50,14 @@ Global option:
 option (atlas_validate.file).allow_unknown_fields = false;
 ```
 
+Field option:
+```
+message User {
+   int64  id    = 1 [(atlas_validate.field).read_only = true];
```
Comment (Evgeniy-L) [inline README.md]: inconsistent with example.proto

Comment (burov) [reply README.md]: Done

**Comment by burov** [comment]
@askurydzin @Evgeniy-L Please review new options design

File: plugin/plugin.go [hunk 2]
```diff
@@ -212,7 +212,7 @@ func (p *Plugin) renderValidatorObjectMethods() {
 
 func (p *Plugin) renderValidatorObjectMethod(o *descriptor.DescriptorProto, t string) {
 	p.P(`// validate_Object_`, t, ` function validates a JSON for a given object.`)
-	p.P(`func validate_Object_`, t, `(r json.RawMessage, path string, allowUnknown bool) (err error) {`)
+	p.P(`func validate_Object_`, t, `(ctx context.Context,r json.RawMessage, path string, allowUnknown bool) (err error) {`)
 	p.P(`obj := &`, t, `{}`)
 	p.P(`if hook, ok := interface{}(obj).(interface { AtlasJSONValidate(json.RawMessage, string, bool) (json.RawMessage, error) }); ok {`)
```
Comment (askurydzin) [inline plugin/plugin.go]: context missing

File: plugin/plugin.go [hunk 3]
```diff
@@ -225,13 +225,37 @@ func (p *Plugin) renderValidatorObjectMethod(o *descriptor.DescriptorProto, t st
...
 		}
+		if fExt, err := proto.GetExtension(f.Options, av_opts.E_Field); err == nil && fExt != nil {
+			favOpt := fExt.(*av_opts.AtlasValidateFieldOption)
+			if favOpt.ReadOnly {
+				p.P(`method, _ := ctx.Value("http-method").(string)`)
+				p.P(`if method != "GET" {`)
+				p.P(`return fmt.Errorf("Field %s has readonly access", k)`)
+				p.P(`}`)
+			} else if methods := GetDenyOperations(favOpt.GetDeny()); len(methods) != 0 {
+				p.P(`method, _ := ctx.Value("http-method").(string)`)
```
Comment (askurydzin) [inline plugin/plugin.go]: better to add HTTPMethodFromCtx (or so) in runtime library to avoid this casting mess.

File: plugin/plugin.go [hunk 4]
```diff
@@ -225,13 +225,37 @@ func (p *Plugin) renderValidatorObjectMethod(o *descriptor.DescriptorProto, t st
...
+				p.WriteString("if v[k] != nil && (")
```
Comment (askurydzin) [inline plugin/plugin.go]: v[k] can be 'null', it must be non-present in v, which is by default false. k is always in v.

File: plugin/plugin.go [hunk 5]
```diff
@@ -225,13 +225,37 @@ func (p *Plugin) renderValidatorObjectMethod(o *descriptor.DescriptorProto, t st
...
+				for ind, m := range methods {
```
Comment (askurydzin) [inline plugin/plugin.go]: strings.Join might be very helpful.

File: plugin/plugin.go [hunk 6]
```diff
@@ -225,13 +225,37 @@ func (p *Plugin) renderValidatorObjectMethod(o *descriptor.DescriptorProto, t st
...
+					p.WriteString(fmt.Sprintf(`method == "%s"`, m))
+					if ind == len(methods)-1 {
+						p.WriteString(") {")
+						break
+					}
+					p.WriteString(" || ")
+				}
+				p.P(`return fmt.Errorf("Field %s has unsupported for method %s", k, method)`)
```
Comment (askurydzin) [inline plugin/plugin.go]: simplier is better, maybe just paste

if /* deny.update = true */ {
p.P(`if method == "POST" {`)
p.P(`return fmt.Errorf("Field %q is not allowed for create method??")`)

etc/

File: plugin/plugin.go (hunk 2 above)
Comment (burov) [reply plugin/plugin.go]: In this pull request i don't break custom validate method contract, in next pull request i move all data to context and add context to custom validate method

File: plugin/plugin.go (hunk 4 above)
Comment (burov) [reply plugin/plugin.go]: null valid value in some cases, on gateway level if we set null token , null convert to default value, in our case only not presented json key valid behavior

File: plugin/plugin.go (hunk 2 above)
Comment (askurydzin) [reply plugin/plugin.go]: ok

File: plugin/plugin.go (hunk 4 above)
Comment (burov) [reply plugin/plugin.go]: done

File: plugin/plugin.go [hunk 7]
```diff
@@ -225,13 +225,33 @@ func (p *Plugin) renderValidatorObjectMethod(o *descriptor.DescriptorProto, t st
...
 	for _, f := range o.GetField() {
 		p.P(`case "`, f.GetName(), `":`)
 		gt := p.getGoType(f.GetTypeName())
 		if p.IsMap(f) {
 			continue
 		}
+		if fExt, err := proto.GetExtension(f.Options, av_opts.E_Field); err == nil && fExt != nil {
+			favOpt := fExt.(*av_opts.AtlasValidateFieldOption)
+			if favOpt.ReadOnly {
+				p.P(`if validate_runtime.HTTPMethodFromContext(ctx) != "GET" {`)
```
Comment (askurydzin) [inline plugin/plugin.go]: not sure what does this mean, body for GET request is always empty.

Comment (askurydzin) [reply plugin/plugin.go]: so, just return error if favOpt.ReadOnly

**Comment by burov** [comment]
@Evgeniy-L  Please do final review

**Pull request merged**